In [0]:
%pip install torch torchvision torchaudio


Note: you may need to restart the kernel using %restart_python or dbutils.library.restartPython() to use updated packages.


In [0]:
import torch
print(torch.__version__)
print("Torch installed correctly")

2.9.1+cpu
Torch installed correctly


In [0]:
%pip install --upgrade --force-reinstall \
  datasets==2.18.0 \
  huggingface_hub==0.20.3 \
  fsspec==2023.10.0

%restart_python


INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 56.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.4/14.4 MB 122.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.0/45.0 MB 190.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 775.1/775.1 kB 28.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 144.6 MB/s eta 0:00:00
  Attempting uninstall: pytz
    Found existing installation: pytz 2025.2
    Uninstalling pytz-2025.2:
      Successfully uninstalled pytz-2025.2
  Attempting uninstall: xxhash
    Found existing installation: xxhash 3.6.0
    Uninstalling xxhash-3.6.0:
      Successfully uninstalled xxhash-3.6.0
  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.6.2
    Uninstalling urllib3-2.6.2:
      Successfully uninstalled urllib3-2.6.

In [0]:
# import datasets
# import huggingface_hub
# import fsspec

# print("datasets:", datasets.__version__)
# print("huggingface_hub:", huggingface_hub.__version__)
# print("fsspec:", fsspec.__version__)


# from datasets import load_dataset
# print("datasets installed correctly")

# dataset = load_dataset("AIBIC/MLOmics") // can't load entire dataset because AIBIC/MLOmics Hugging Face dataset is NOT a single uniform table
# dataset

from huggingface_hub import list_repo_files

files = list_repo_files("AIBIC/MLOmics", repo_type="dataset")

print("Total files:", len(files))

brca_files = [f for f in files if "BRCA" in f]
len(brca_files), brca_files




Total files: 242


(19,
 ['Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_CNV_aligned.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_Methy_aligned.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_label_num.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_mRNA_aligned.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_miRNA_aligned.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_CNV.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_Methy.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_label_num.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_mRNA.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_miRNA.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_CNV_top.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_Methy_top.csv',
  'Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_label_num.csv',
  'Main_D

In [0]:
from collections import defaultdict

grouped = defaultdict(list)

for f in brca_files:
    if "Classification" in f:
        grouped["classification"].append(f)
    elif "Clustering" in f:
        grouped["clustering"].append(f)
    elif "Imputation" in f:
        grouped["imputation"].append(f)
    elif "Survival" in f:
        grouped["survival"].append(f)
    else:
        grouped["other"].append(f)

for k, v in grouped.items():
    print(f"\n{k.upper()} ({len(v)} files)")
    for x in v:
        print(" ", x)


CLASSIFICATION (15 files)
  Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_CNV_aligned.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_Methy_aligned.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_label_num.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_mRNA_aligned.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_miRNA_aligned.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_CNV.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_Methy.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_label_num.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_mRNA.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_miRNA.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_CNV_top.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_Methy_top.csv
  Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_label_num.csv
  Main_Dataset/Classificat

In [0]:
from huggingface_hub import hf_hub_download

repo_id = "AIBIC/MLOmics"

# Test 1 file download
cnv_path = hf_hub_download(
    repo_id=repo_id,
    filename="Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_CNV.csv",
    repo_type="dataset"
)

cnv_path


BRCA_CNV.csv:   0%|          | 0.00/16.6M [00:00<?, ?B/s]

'/home/spark-823d78bc-2848-4ce5-a7f7-e9/.cache/huggingface/hub/datasets--AIBIC--MLOmics/snapshots/d7c2239504dcf817ee9f33a693f9617d614eef67/Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_CNV.csv'

In [0]:
# import pandas as pd

# cnv = pd.read_csv(
#     "/home/spark-74dee5de-088a-43f8-8744-95/.cache/huggingface/hub/datasets--AIBIC--MLOmics/snapshots/d7c2239504dcf817ee9f33a693f9617d614eef67/Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_CNV.csv",
#     index_col=0
# )

# cnv.shape


# # Confirming patient id's look right
# cnv.columns[:5]



In [0]:
from huggingface_hub import hf_hub_download
import pandas as pd

repo_id = "AIBIC/MLOmics"

files = {
    "CNV": "Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_CNV.csv",
    "Methy": "Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_Methy.csv",
    "mRNA": "Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_mRNA.csv",
    "miRNA": "Main_Dataset/Imputation_datasets/Imp-BRCA/Top/BRCA_miRNA.csv"
}

dfs = {}

for name, filename in files.items():
    path = hf_hub_download(
        repo_id=repo_id,
        filename=filename,
        repo_type="dataset"    
    )
    df = pd.read_csv(path, index_col=0)
    dfs[name] = df
    print(f"{name} shape: ", df.shape)

CNV shape:  (2000, 671)


BRCA_Methy.csv:   0%|          | 0.00/16.6M [00:00<?, ?B/s]

Methy shape:  (2000, 671)


BRCA_mRNA.csv:   0%|          | 0.00/16.7M [00:00<?, ?B/s]

mRNA shape:  (2000, 671)


BRCA_miRNA.csv:   0%|          | 0.00/1.68M [00:00<?, ?B/s]

miRNA shape:  (200, 671)


In [0]:
# Compute patient overlap
patient_sets = [set(df.columns) for df in dfs.values()]
common_patients = set.intersection(*patient_sets)

print("Patients will all 4 omics:", len(common_patients)) # Patients will all 4 omics: 671

Patients will all 4 omics: 671


In [0]:
common_patients = (
    set(dfs["CNV"].columns)
    & set(dfs["Methy"].columns)
    & set(dfs["mRNA"].columns)
    & set(dfs["miRNA"].columns)
)

common_patients = sorted(common_patients)

for name in dfs:
    print(name, dfs[name].shape)


print("Patients with all 4 omics:", len(common_patients))

# Output (using imputation datasets --> imputation reconstructs data):
# CNV (2000, 671)
# Methy (2000, 671)
# mRNA (2000, 671)
# miRNA (200, 671)
# Patients with all 4 omics: 671

CNV (2000, 671)
Methy (2000, 671)
mRNA (2000, 671)
miRNA (200, 671)
Patients with all 4 omics: 671


In [0]:
# Task 1 BRCA subtype classification
# Labels:
#   0 = Luminal A
#   1 = Luminal B
#   2 = HER2-enriched
#   3 = Basal-like


# MLOmics classification Structure
# Main_Dataset/
# └── Classification_datasets/
#     └── GS-BRCA/
#         └── Top/
#             ├── BRCA_mRNA.csv --> genes × patients --> Gene expression
#             ├── BRCA_CNV.csv --> regions × patients --> Copy number
#             ├── BRCA_Methy.csv --> CpGs × patients --> DNA methylation
#             ├── BRCA_miRNA.csv --> miRNAs × patients --> miRNA expression
#             └── BRCA_label_num.csv --> patients × 1 --> ground truth labels


# Important: columns are patients, rows are features

In [0]:
# Switching to Classification_datasets/GS-BRCA to predict clinical labels

from huggingface_hub import HfApi

api = HfApi()

files = api.list_repo_files(
    repo_id="AIBIC/MLOmics",
    repo_type="dataset"
)

# Show only GS-BRCA files
gs_brca_files = [f for f in files if "Classification_datasets/GS-BRCA" in f]

for f in gs_brca_files:
    print(f)


# Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_CNV_aligned.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_Methy_aligned.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_label_num.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_mRNA_aligned.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_miRNA_aligned.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_CNV.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_Methy.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_label_num.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_mRNA.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_miRNA.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_CNV_top.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_Methy_top.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_label_num.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_mRNA_top.csv
# Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_miRNA_top.csv


# 1️⃣ Original

#   Raw, very high dimensional

#   Harder to model

#   Useful for deep learning later

# 2️⃣ Aligned

#   Features aligned across omics

#   Good for representation learning

# 3️⃣ Top ✅ (what we want now)

#   Preselected top features

#   Lower dimensional

#   Designed for classical ML baselines

#   Best starting point


# will be using Main_Dataset/Classification_datasets/GS-BRCA/Top/:
# BRCA_mRNA_top.csv, BRCA_CNV_top.csv, BRCA_Methy_top.csv, BRCA_miRNA_top.csv, BRCA_label_num.csv


Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_CNV_aligned.csv
Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_Methy_aligned.csv
Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_label_num.csv
Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_mRNA_aligned.csv
Main_Dataset/Classification_datasets/GS-BRCA/Aligned/BRCA_miRNA_aligned.csv
Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_CNV.csv
Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_Methy.csv
Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_label_num.csv
Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_mRNA.csv
Main_Dataset/Classification_datasets/GS-BRCA/Original/BRCA_miRNA.csv
Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_CNV_top.csv
Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_Methy_top.csv
Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_label_num.csv
Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_mRNA_top.csv
Main_Dataset

In [0]:
from huggingface_hub import hf_hub_download
import pandas as pd

repo_id = "AIBIC/MLOmics"

files = {
    "CNV":   "Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_CNV_top.csv",
    "Methy": "Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_Methy_top.csv",
    "mRNA":  "Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_mRNA_top.csv",
    "miRNA": "Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_miRNA_top.csv",
    "Label": "Main_Dataset/Classification_datasets/GS-BRCA/Top/BRCA_label_num.csv",
}

dfs_cls = {}

for name, filename in files.items():
    path = hf_hub_download(
        repo_id=repo_id,
        filename=filename,
        repo_type="dataset"
    )
    df = pd.read_csv(path, index_col=0)
    dfs_cls[name] = df
    print(f"{name} shape: {df.shape}")


labels = dfs_cls["Label"]

print(labels.head())
print(labels.index[:5])
print(labels.shape)

# *** DO Not align labels by patient id, MUST lbael by column order


# Define common patients using omics only
common_patients = (
    set(dfs_cls["CNV"].columns)
    & set(dfs_cls["Methy"].columns)
    & set(dfs_cls["mRNA"].columns)
    & set(dfs_cls["miRNA"].columns)
    # & set(dfs_cls["Label"].index)
)

common_patients = sorted(common_patients)
# print("Patients with all omics:", len(common_patients))

BRCA_CNV_top.csv:   0%|          | 0.00/61.4M [00:00<?, ?B/s]

CNV shape: (5000, 671)


BRCA_Methy_top.csv:   0%|          | 0.00/61.6M [00:00<?, ?B/s]

Methy shape: (5000, 671)


BRCA_mRNA_top.csv:   0%|          | 0.00/61.3M [00:00<?, ?B/s]

mRNA shape: (5000, 671)


BRCA_miRNA_top.csv:   0%|          | 0.00/1.79M [00:00<?, ?B/s]

miRNA shape: (366, 671)


BRCA_label_num.csv:   0%|          | 0.00/1.35k [00:00<?, ?B/s]

Label shape: (671, 0)
Empty DataFrame
Columns: []
Index: [0, 1, 2, 0, 0]
Index([0, 1, 2, 0, 0], dtype='int64', name='Label')
(671, 0)


In [0]:
# Subset omics in fixed order
X_mrna = dfs_cls["mRNA"][common_patients].T
X_cnv = dfs_cls["CNV"][common_patients].T
X_methy = dfs_cls["Methy"][common_patients].T
X_mirna = dfs_cls["miRNA"][common_patients].T
y = dfs_cls["Label"].index.values

print(X_mrna.shape)
print(X_cnv.shape)
print(X_methy.shape)
print(X_mirna.shape)
print(y.shape)
print(set(y))

assert X_mrna.shape[0] == len(y)


# Main takeaway:
#   * 671 patients
#   * 4 omics modalities, fully aligned
#   * Top feature sets
#   * 5-class classification problem

# mRNA   : (671, 5000)
# CNV    : (671, 5000)
# Methy  : (671, 5000)
# miRNA  : (671, 366)
# y      : (671,)
# Classes: {0,1,2,3,4}


(671, 5000)
(671, 5000)
(671, 5000)
(671, 366)
(671,)
{np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4)}


In [0]:
# Starting with single omics baseline, starting with mrna only first, then cnv, methy, mirna

# Check class balance: tells how hard the problem is
import numpy as np

unique, counts = np.unique(y, return_counts=True)
for cls, cnt in zip(unique, counts):
    print(f"Class {cls}: {cnt} samples")


# Class 0: 353 samples
# Class 1: 42 samples
# Class 2: 132 samples
# Class 3: 31 samples
# Class 4: 113 samples


Class 0: 353 samples
Class 1: 42 samples
Class 2: 132 samples
Class 3: 31 samples
Class 4: 113 samples


In [0]:
# Train mRNA baseline classifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, classification_report

X_mrna_train, X_mrna_test, y_mrna_train, y_mrna_test = train_test_split(X_mrna, y, test_size=0.2, stratify=y, random_state=42)

mrna_clf = LogisticRegression(max_iter=1000, n_jobs=-1, multi_class="auto")

mrna_clf.fit(X_mrna_train, y_mrna_train)

y_mrna_pred = mrna_clf.predict(X_mrna_test)

print("mRNA Accuracy:", accuracy_score(y_mrna_test, y_mrna_pred))
print("mRNA Macro F1:", f1_score(y_mrna_test, y_mrna_pred, average="macro"))
print("\nClassification report:\n")
print(classification_report(y_mrna_test, y_mrna_pred))


# mRNA Accuracy: 0.8518518518518519
# mRNA Macro F1: 0.7721375930144718

# Classification report:

#               precision    recall  f1-score   support

#            0       0.88      0.93      0.90        71
#            1       0.83      0.62      0.71         8
#            2       0.82      0.67      0.73        27
#            3       0.50      0.67      0.57         6
#            4       0.92      0.96      0.94        23

#     accuracy                           0.85       135
#    macro avg       0.79      0.77      0.77       135
# weighted avg       0.85      0.85      0.85       135



/databricks/python/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
I0000 00:00:1766274259.315960    3453 fork_posix.cc:77] Other threads are currently calling into gRPC, skipping fork() handlers
I0000 00:00:1766274259.365408    3453 fork_posix.cc:77] Other threads are currently calling into gRPC, skipping fork() handlers
I0000 00:00:1766274259.382714    3453 fork_posix.cc:77] Other threads are currently calling into gRPC, skipping fork() handlers


mRNA Accuracy: 0.8518518518518519
mRNA Macro F1: 0.7721375930144718

Classification report:

              precision    recall  f1-score   support

           0       0.88      0.93      0.90        71
           1       0.83      0.62      0.71         8
           2       0.82      0.67      0.73        27
           3       0.50      0.67      0.57         6
           4       0.92      0.96      0.94        23

    accuracy                           0.85       135
   macro avg       0.79      0.77      0.77       135
weighted avg       0.85      0.85      0.85       135



In [0]:
# Logging model using MLFlow for mRNA baseline

import mlflow
import mlflow.sklearn
from sklearn.metrics import accuracy_score, f1_score

input_example = X_mrna_train.iloc[:5]

with mlflow.start_run(run_name="mRNA_baseline_v1"):
    mlflow.log_param("omics", "mRNA")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("model", "LogisticRegression")
    mlflow.log_param("scaling", "standard")

    mlflow.log_metric("accuracy", accuracy_score(y_mrna_test, y_mrna_pred))
    mlflow.log_metric("macro_f1", f1_score(y_mrna_test, y_mrna_pred, average="macro"))

    mlflow.sklearn.log_model(
        mrna_clf,
        artifact_path="model",
        input_example=input_example
    )


/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


Uploading artifacts:   0%|          | 0/11 [00:00<?, ?it/s]

In [0]:
# Train methylation baseline classifier
X_methy_train, X_methy_test, y_methy_train, y_methy_test = train_test_split(
    X_methy,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

methy_clf = LogisticRegression(
    max_iter=3000,
    n_jobs=-1,
    multi_class="auto"
)

methy_clf.fit(X_methy_train, y_methy_train)
y_methy_pred = methy_clf.predict(X_methy_test)

print("Methy Accuracy:", accuracy_score(y_methy_test, y_methy_pred))
print("Methy Macro F1:", f1_score(y_methy_test, y_methy_pred, average="macro"))
print("\nClassification report:\n")
print(classification_report(y_methy_test, y_methy_pred))


# Methy Accuracy: 0.7777777777777778
# Methy Macro F1: 0.7026730692276912

# Classification report:

#               precision    recall  f1-score   support

#            0       0.82      0.87      0.84        71
#            1       0.67      0.75      0.71         8
#            2       0.59      0.48      0.53        27
#            3       0.50      0.50      0.50         6
#            4       0.95      0.91      0.93        23

#     accuracy                           0.78       135
#    macro avg       0.71      0.70      0.70       135
# weighted avg       0.77      0.78      0.77       135


/databricks/python/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


Methy Accuracy: 0.7777777777777778
Methy Macro F1: 0.7026730692276912

Classification report:

              precision    recall  f1-score   support

           0       0.82      0.87      0.84        71
           1       0.67      0.75      0.71         8
           2       0.59      0.48      0.53        27
           3       0.50      0.50      0.50         6
           4       0.95      0.91      0.93        23

    accuracy                           0.78       135
   macro avg       0.71      0.70      0.70       135
weighted avg       0.77      0.78      0.77       135



In [0]:
# Logging model using MLFlow for methylation baseline

input_example = X_methy_train.iloc[:5]

with mlflow.start_run(run_name="Methy_baseline_v1"):
    mlflow.log_param("omics", "Methy")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("model", "LogisticRegression")
    mlflow.log_param("scaling", "standard")

    mlflow.log_metric("accuracy", accuracy_score(y_methy_test, y_methy_pred))
    mlflow.log_metric("macro_f1", f1_score(y_methy_test, y_methy_pred, average="macro"))

    mlflow.sklearn.log_model(
        methy_clf,
        artifact_path="model",
        input_example=input_example
    )

/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


Uploading artifacts:   0%|          | 0/11 [00:00<?, ?it/s]

In [0]:
# Train cnv baseline classifier
X_cnv_train, X_cnv_test, y_cnv_train, y_cnv_test = train_test_split(
    X_cnv,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

cnv_clf = LogisticRegression(
    max_iter=3000,
    n_jobs=-1,
    random_state=42
)

cnv_clf.fit(X_cnv_train, y_cnv_train)
y_cnv_pred = cnv_clf.predict(X_cnv_test)
print("CNV Accuracy:", accuracy_score(y_cnv_test, y_cnv_pred))
print("CNV Macro F1:", f1_score(y_cnv_test, y_cnv_pred, average="macro"))
print("\nClassification report:\n")
print(classification_report(y_cnv_test, y_cnv_pred))


# CNV Accuracy: 0.6962962962962963
# CNV Macro F1: 0.5596961916987592

# Classification report:

#               precision    recall  f1-score   support

#            0       0.75      0.86      0.80        71
#            1       0.36      0.50      0.42         8
#            2       0.50      0.26      0.34        27
#            3       0.50      0.33      0.40         6
#            4       0.80      0.87      0.83        23

#     accuracy                           0.70       135
#    macro avg       0.58      0.56      0.56       135
# weighted avg       0.68      0.70      0.68       135



CNV Accuracy: 0.6962962962962963
CNV Macro F1: 0.5596961916987592

Classification report:

              precision    recall  f1-score   support

           0       0.75      0.86      0.80        71
           1       0.36      0.50      0.42         8
           2       0.50      0.26      0.34        27
           3       0.50      0.33      0.40         6
           4       0.80      0.87      0.83        23

    accuracy                           0.70       135
   macro avg       0.58      0.56      0.56       135
weighted avg       0.68      0.70      0.68       135



In [0]:
# Logging model using MLFlow for CNV baseline

input_example = X_cnv_train.iloc[:5]

with mlflow.start_run(run_name="CNV_baseline_v1"):
    mlflow.log_param("omics", "CNV")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("model", "LogisticRegression")
    mlflow.log_param("scaling", "standard")

    mlflow.log_metric("accuracy", accuracy_score(y_cnv_test, y_cnv_pred))
    mlflow.log_metric("macro_f1", f1_score(y_cnv_test, y_cnv_pred, average="macro"))

    mlflow.sklearn.log_model(
        cnv_clf,
        artifact_path="model",
        input_example=input_example
    )

/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


Uploading artifacts:   0%|          | 0/11 [00:00<?, ?it/s]

In [0]:
# Train mirna baseline classifier
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer


original_columns = X_mirna.columns.copy()
X_mirna.columns = X_mirna.columns.astype(str)

# X_mirna.columns = original_columns

X_mirna_clean = X_mirna.dropna(axis=1, how="all")

print("Original miRNA features:", X_mirna.shape[1])
print("After dropping all-NaN features:", X_mirna_clean.shape[1])

X_mirna_train, X_mirna_test, y_mirna_train, y_mirna_test = train_test_split(
    X_mirna_clean,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

# pipeline = Pipeline([
#     ("imputer", SimpleImputer(strategy="median")),
#     ("clf", LogisticRegression(
#         max_iter=3000,
#         n_jobs=-1,
#         multi_class="auto"
#     ))
# ])

# Fraction of missing miRNA values: 0.454
# nan_fraction = X_mirna.isna().mean().mean()
# print(f"Fraction of missing miRNA values: {nan_fraction:.3f}")


# pipeline.fit(X_train, y_train)
# y_pred = pipeline.predict(X_test)


# Confirming that miRNA limitation is not model-specific

# Using this model as miRNA baseline: miRNA-only baseline (after sparsity-aware preprocessing)

mirna_clf = HistGradientBoostingClassifier(random_state=42)
mirna_clf.fit(X_mirna_train, y_mirna_train)
y_mirna_pred = mirna_clf.predict(X_mirna_test)

print("miRNA Accuracy:", accuracy_score(y_mirna_test, y_mirna_pred))
print("miRNA Macro F1:", f1_score(y_mirna_test, y_mirna_pred, average="macro"))
print("\nClassification report:\n")
print(classification_report(y_mirna_test, y_mirna_pred))



# miRNA Accuracy: 0.8074074074074075
# miRNA Macro F1: 0.6952245931555565

# Classification report:

#               precision    recall  f1-score   support

#            0       0.86      0.94      0.90        71
#            1       0.44      0.50      0.47         8
#            2       0.72      0.48      0.58        27
#            3       0.57      0.67      0.62         6
#            4       0.91      0.91      0.91        23

#     accuracy                           0.81       135
#    macro avg       0.70      0.70      0.70       135
# weighted avg       0.80      0.81      0.80       135


# print("miRNA Accuracy:", accuracy_score(y_test, y_pred))
# print("miRNA Macro F1:", f1_score(y_test, y_pred, average="macro"))
# print("\nClassification report:\n")
# print(classification_report(y_test, y_pred)) 

Original miRNA features: 366
After dropping all-NaN features: 200
miRNA Accuracy: 0.8074074074074075
miRNA Macro F1: 0.6952245931555565

Classification report:

              precision    recall  f1-score   support

           0       0.86      0.94      0.90        71
           1       0.44      0.50      0.47         8
           2       0.72      0.48      0.58        27
           3       0.57      0.67      0.62         6
           4       0.91      0.91      0.91        23

    accuracy                           0.81       135
   macro avg       0.70      0.70      0.70       135
weighted avg       0.80      0.81      0.80       135



In [0]:
# Logging model using MLFlow for miRNA baseline

input_example = X_mirna_train.iloc[:5]

with mlflow.start_run(run_name="miRNA_baseline_v1"):
    mlflow.log_param("omics", "miRNA")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("model", "LogisticRegression")
    mlflow.log_param("scaling", "standard")

    mlflow.log_metric("accuracy", accuracy_score(y_mirna_test, y_mirna_pred))
    mlflow.log_metric("macro_f1", f1_score(y_mirna_test, y_mirna_pred, average="macro"))

    mlflow.sklearn.log_model(
        mirna_clf,
        artifact_path="model",
        input_example=input_example
    )

/databricks/python/lib/python3.12/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


Uploading artifacts:   0%|          | 0/11 [00:00<?, ?it/s]

In [0]:
# Sanity check:
assert X_cnv.index.equals(X_mrna.index)
assert X_cnv.index.equals(X_methy.index)
assert X_cnv.index.equals(X_mirna.index)


In [0]:
# EARLY FUSION (CONCATENATE FEATURES)
import pandas as pd

X_fusion = pd.concat([X_cnv, X_mrna, X_methy, X_mirna], axis=1)

print("Fusion shape:", X_fusion.shape)

Fusion shape: (671, 15366)


In [0]:
# Train early fusion model
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, classification_report

X_fusion_clean = X_fusion.dropna(axis=1, how="all")

print("Before:", X_fusion.shape)
print("After:", X_fusion_clean.shape)


Xf_train, Xf_test, yf_train, yf_test = train_test_split(
    X_fusion_clean,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

fusion_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler()),
    ("clf", LogisticRegression(
        solver="saga",
        max_iter=3000,
        n_jobs=-1,
        random_state=42
    ))
])

fusion_pipeline.fit(Xf_train, yf_train)
yf_pred = fusion_pipeline.predict(Xf_test)

print("Fusion Accuracy:", accuracy_score(yf_test, yf_pred))
print("Fusion Macro F1:", f1_score(yf_test, yf_pred, average="macro"))
print("Fusion Classification Report:")
print(classification_report(yf_test, yf_pred))


# Fusion Accuracy: 0.8296296296296296
# Fusion Macro F1: 0.7374042248478339
# Fusion Classification Report:
#               precision    recall  f1-score   support

#            0       0.93      0.90      0.91        71
#            1       0.55      0.75      0.63         8
#            2       0.73      0.59      0.65        27
#            3       0.50      0.67      0.57         6
#            4       0.88      0.96      0.92        23

#     accuracy                           0.83       135
#    macro avg       0.72      0.77      0.74       135
# weighted avg       0.84      0.83      0.83       135



# Key insight: *** Multi-omics integration improves robustness and subtype sensitivity rather than raw accuracy. ***


Before: (671, 15366)
After: (671, 15200)
Fusion Accuracy: 0.8296296296296296
Fusion Macro F1: 0.7374042248478339
Fusion Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.90      0.91        71
           1       0.55      0.75      0.63         8
           2       0.73      0.59      0.65        27
           3       0.50      0.67      0.57         6
           4       0.88      0.96      0.92        23

    accuracy                           0.83       135
   macro avg       0.72      0.77      0.74       135
weighted avg       0.84      0.83      0.83       135



In [0]:
# Log to MLFlow
import mlflow
import mlflow.sklearn

with mlflow.start_run(run_name="EarlyFusion_baseline_v1"):
    mlflow.log_param("fusion_type", "early")
    mlflow.log_param("omics", "CNV+mRNA+Methy+miRNA")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("model", "LogisticRegression")
    mlflow.log_param("n_features", Xf_train.shape[1])

    mlflow.log_metric("accuracy", accuracy_score(yf_test, yf_pred))
    mlflow.log_metric("macro_f1", f1_score(yf_test, yf_pred, average="macro"))

    mlflow.sklearn.log_model(
        fusion_pipeline,
        artifact_path="model",
        input_example=Xf_train.iloc[:5]
    )


/databricks/python/lib/python3.12/site-packages/mlflow/models/utils.py:217: UserWarning: DataFrame columns are not unique, some columns will be omitted.
  result = _handle_dataframe_nans(df).to_dict(orient="split")
2025/12/20 23:49:18 WARNING mlflow.models.signature: Failed to infer the model signature from the input example. Reason: TypeError("Expected pandas.Series, got '<class 'pandas.core.frame.DataFrame'>'."). To see the full traceback, set the logging level to DEBUG via `logging.getLogger("mlflow").setLevel(logging.DEBUG)`.
2025/12/20 23:49:22 WARNING mlflow.models.model: Model logged without a signature. Signatures are required for Databricks UC model registry as they validate model inputs and denote the expected schema of model outputs. Please visit https://www.mlflow.org/docs/2.22.0/model/signatures.html#how-to-set-signatures-on-models for instructions on setting signature on models.


Uploading artifacts:   0%|          | 0/11 [00:00<?, ?it/s]

In [0]:
# Late fusion model
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, classification_report

proba_cnv = cnv_clf.predict_proba(X_cnv_test)
proba_mrna  = mrna_clf.predict_proba(X_mrna_test)
proba_methy = methy_clf.predict_proba(X_methy_test)
proba_mirna = mirna_clf.predict_proba(X_mirna_test)


# weights = {
#     "mrna": 0.35,
#     "methy": 0.25,
#     "mirna": 0.25,
#     "cnv": 0.15
# }

# proba_fused = (
#     weights["mrna"]  * proba_mrna +
#     weights["methy"] * proba_methy +
#     weights["mirna"] * proba_mirna +
#     weights["cnv"]   * proba_cnv
# )

# y_pred = np.argmax(proba_fused, axis=1)


# proba_val = {
#     "cnv": proba_cnv,
#     "mrna": proba_mrna,
#     "methy": proba_methy,
#     "mirna": proba_mirna
# }

# best_f1 = 0
# best_weights = None

# weights_range = np.arange(0.05, 0.61, 0.05)

# for w_mrna in weights_range:
#     for w_methy in weights_range:
#         for w_mirna in weights_range:
#             w_cnv = 1 - (w_mrna + w_methy + w_mirna)
#             if w_cnv < 0:
#                 continue

#             fused_proba = (
#                 w_mrna  * proba_val["mrna"] +
#                 w_methy * proba_val["methy"] +
#                 w_mirna * proba_val["mirna"] +
#                 w_cnv   * proba_val["cnv"]
#             )

#             y_pred = np.argmax(fused_proba, axis=1)
#             f1 = f1_score(yf_test, yf_pred, average="macro")

#             if f1 > best_f1:
#                 best_f1 = f1
#                 best_weights = {
#                     "mrna": w_mrna,
#                     "methy": w_methy,
#                     "mirna": w_mirna,
#                     "cnv": w_cnv
#                 }


# print("Best weights:", best_weights)
# print("Best validation Macro F1:", best_f1)


# w = best_weights

# fused_test_proba = (
#     w["mrna"]  * proba_val["mrna"] +
#     w["methy"] * proba_val["methy"] +
#     w["mirna"] * proba_val["mirna"] +
#     w["cnv"]   * proba_val["cnv"]
# )

# y_test_pred = np.argmax(fused_test_proba, axis=1)

# print("Weighted Late Fusion Accuracy:", accuracy_score(yf_test, y_test_pred))
# print("Weighted Late Fusion Macro F1:", f1_score(yf_test, y_test_pred, average="macro"))
# print(classification_report(yf_test, y_test_pred))



# # print(proba_cnv)
# # print(proba_mrna)
# # print(proba_methy)
# # print(proba_mirna)

# # print(proba_cnv.shape) # (135, 5)
# # print(proba_mrna.shape)
# # print(proba_methy.shape)
# # print(proba_mirna.shape)


# proba_fusion = (proba_cnv + proba_mrna + proba_methy + proba_mirna)/4

# y_pred_fusion = np.argmax(proba_fusion, axis=1)


print("Late Fusion Accuracy:", accuracy_score(yf_test, y_pred_fusion))
print("Late Fusion Macro F1:", f1_score(yf_test, y_pred_fusion, average="macro"))
print("\nLate Fusion Classification Report:\n")
print(classification_report(yf_test, y_pred_fusion))

# Unweighted Late Fusion
# Late Fusion Accuracy: 0.8518518518518519
# Late Fusion Macro F1: 0.7945263897250652

# Late Fusion Classification Report:

#               precision    recall  f1-score   support

#            0       0.85      0.96      0.90        71
#            1       0.75      0.75      0.75         8
#            2       0.82      0.52      0.64        27
#            3       0.80      0.67      0.73         6
#            4       0.92      1.00      0.96        23

#     accuracy                           0.85       135
#    macro avg       0.83      0.78      0.79       135
# weighted avg       0.85      0.85      0.84       135



# Weighted Late Fusion
# Best weights: {'mrna': np.float64(0.05), 'methy': np.float64(0.05), 'mirna': np.float64(0.05), 'cnv': np.float64(0.85)}
# Best validation Macro F1: 0.7374042248478339
# Weighted Late Fusion Accuracy: 0.6962962962962963
# Weighted Late Fusion Macro F1: 0.5596961916987592
#               precision    recall  f1-score   support

#            0       0.75      0.86      0.80        71
#            1       0.36      0.50      0.42         8
#            2       0.50      0.26      0.34        27
#            3       0.50      0.33      0.40         6
#            4       0.80      0.87      0.83        23

#     accuracy                           0.70       135
#    macro avg       0.58      0.56      0.56       135
# weighted avg       0.68      0.70      0.68       135




---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
File <command-6559956291031383>, line 104
      8 proba_mirna = mirna_clf.predict_proba(X_mirna_test)
     11 # weights = {
     12 #     "mrna": 0.35,
     13 #     "methy": 0.25,
   (...)
    100 
    101 # y_pred_fusion = np.argmax(proba_fusion, axis=1)
--> 104 print("Late Fusion Accuracy:", accuracy_score(yf_test, y_pred_fusion))
    105 print("Late Fusion Macro F1:", f1_score(yf_test, y_pred_fusion, average="macro"))
    106 print("\nLate Fusion Classification Report:\n")

NameError: name 'y_pred_fusion' is not defined

Despite individual modalities exhibiting varying predictive strengths, uniform late fusion outperformed optimized weighted fusion, suggesting that equal weighting provides better regularization and robustness under limited sample sizes.
Late fusion consistently outperformed early fusion and individual omics models. Weight optimization did not significantly improve performance over uniform averaging, suggesting that equal-weight late fusion provides a robust and well-regularized integration strategy under limited sample sizes.

In [0]:
import mlflow

with mlflow.start_run(run_name="LateFusion_baseline_v1"):
    mlflow.log_param("fusion_type", "late")
    mlflow.log_param("omics", "CNV+mRNA+Methy+miRNA")
    mlflow.log_param("dataset", "MLOmics GS-BRCA Top")
    mlflow.log_param("fusion_rule", "mean_probability")

    mlflow.log_metric("accuracy", accuracy_score(yf_test, y_pred_fusion))
    mlflow.log_metric("macro_f1", f1_score(yf_test, y_pred_fusion, average="macro"))


## Key report:
Among single-omics baselines, the mRNA classifier achieved the strongest performance (accuracy = 0.852, macro F1 = 0.772). Late fusion of all four omics modalities matched this accuracy while further improving macro F1 to 0.795, primarily due to improved performance on minority classes. This indicates that complementary information from CNV, methylation, and miRNA data enhances balanced classification performance beyond what is achievable with mRNA alone.

In [0]:
# Stacked Late fusion
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd

X_dummy = X_mrna  # just to get indices
y_all = y

idx_train, idx_test = train_test_split(
    X_dummy.index,
    test_size=0.2,
    stratify=y_all,
    random_state=42
)


def train_and_predict_proba(X, y_series, idx_train, idx_test):
    pipeline = Pipeline([
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler()),
        ("clf", LogisticRegression(
            solver="liblinear",
            random_state=42
        ))
    ])

    pipeline.fit(X.loc[idx_train], y_series[idx_train])
    proba_test = pipeline.predict_proba(X.loc[idx_test])

    return pipeline, proba_test

y_series = pd.Series(y, index=X_mrna.index)


models = {}
probas = {}

for name, X in {
    "mrna": X_mrna,
    "methy": X_methy,
    "mirna": X_mirna,
    "cnv": X_cnv
}.items():
    model, proba = train_and_predict_proba(
        X, y_series, idx_train, idx_test
    )
    models[name] = model
    probas[name] = proba


X_stack = np.hstack([
    probas["mrna"],
    probas["methy"],
    probas["mirna"],
    probas["cnv"]
])

# X_stack.shape


meta_clf = LogisticRegression(
    solver="lbfgs",
    max_iter=1000,
    random_state=42
)

meta_clf.fit(X_stack, y_series[idx_test])


y_true = y_series.loc[idx_test]
y_pred = meta_clf.predict(X_stack)

print("Stacked Late Fusion Accuracy:", accuracy_score(y_true, y_pred))
print("Stacked Late Fusion Macro F1:", f1_score(y_true, y_pred, average="macro"))
print("\nClassification report:\n")
print(classification_report(y_true, y_pred))





Stacked late fusion does not materially outperform uniform late fusion in macro F1, though it slightly improves accuracy.

In [0]:
import pandas as pd

feature_names = (
    [f"mRNA_{i}" for i in range(5)] +
    [f"Methy_{i}" for i in range(5)] +
    [f"miRNA_{i}" for i in range(5)] +
    [f"CNV_{i}" for i in range(5)]
)

coef_df = pd.DataFrame(
    meta_clf.coef_,
    columns=feature_names
)

coef_df


Across all classes:
  * mRNA coefficients are the largest magnitude
  * Consistently positive for the correct class
  * Negative for competing classes

Transcriptomics is the strongest single modality for BRCA subtype classification. This aligns with TCGA biology and literature.


miRNA is class specific
  Coefficients are small globally, large for specific classes, and have mixed signs. This means miRNA is not a general classifier, but helps specific subtypes. That explains why miRNA alone is weaker. But improves late fusion


Methylation helps itermediate classes:
  Methy coefficients are strong for classes 0 and 4. Less dominant for classes 1 - 3.


CNV contributes the least.